Rewrite camera modes; implement per-player camera modes

ClobberXD opened the pull request:
#### Use-cases

- 3rd-person mode generally offers players unfair advantages over players using 1st-person mode, like peripheral vision (looking around walls), noclip glitch (seeing through nodes in narrow passageways), etc.
- Gun mods work with the assumption that the player is in 1st-person mode, and if the player is in 3rd-person mode, most calculations will be borked and incorrect.

#### Changes

- enum `CameraMode` moved to class `Player`.
- New `EnumString` array `es_CameraModes` to store all camera mode enums and their string representation pairs.
- New attribute `std::set<CameraMode> Player::m_camera_modes` to store the enabled camera modes (both client-side and server-side) along with getter `Player::getCameraModes` and setter `Player::setCameraModes`.
- New Server->client packet (`TOCLIENT_CAMERA_MODES`).
- New Lua API methods `ObjectRef:get/set_camera_modes`.
- New method `Server::SendCameraModes` for serializing `Player::m_camera_modes` into a `u16` and sending it over the network.
- New client-side packet handler `Client::handleCommand_CameraModes` to deserialize the received `u16` back to `std::set<CameraMode>` and store it client-side.
- Documentation!

(Note: This PR also includes a couple of code-style fixes around the touched code)

sofar commented on src/network/clientopcodes.cpp:
```diff
@@ -56,72 +56,72 @@ const ToClientCommandHandler toClientCommandTable[TOCLIENT_NUM_MSG_TYPES] =
+	null_command_handler,
+	null_command_handler,
+	{ "TOCLIENT_SRP_BYTES_S_B",                 TOCLIENT_STATE_NOT_CONNECTED, &Client::handleCommand_SrpBytesSandB }, // 0x60
+	{ "TOCLIENT_FORMSPEC_PREPEND",              TOCLIENT_STATE_CONNECTED, &Client::handleCommand_FormspecPrepend }, // 0x61,
```
Can't see your changes - too many unrelated whitespace changes.

ClobberXD replied:
Sorry about that, I indented all the other lines to make up for the extra large length of the network packet in question. :)

sofar commented on src/network/serveropcodes.cpp:
```diff
@@ -145,70 +145,70 @@ const ClientCommandFactory clientCommandFactoryTable[TOCLIENT_NUM_MSG_TYPES] =
-	{ "TOCLIENT_CSM_RESTRICTION_FLAGS",    0, true }, // 0x2A
+	{ "TOCLIENT_TIME_OF_DAY",                   0, true }, // 0x29
+	{ "TOCLIENT_CSM_RESTRICTION_FLAGS",         0, true }, // 0x2A
+	{ "TOCLIENT_CAMERA_MODE_RESTRICTION_FLAGS", 0, true }, // 0x2B
```
`_MODE_` could be removed, maybe just make it TOCLIENT_CAMERA_RESTRICTION_FLAGS ? Not that important though.

ClobberXD replied:
I felt `mode` made it clear that this restricts camera modes, and not the camera itself (i.e. camera movement, rotation, etc.)

ClobberXD replied:
Should I omit `_MODE_` and if yes, should I omit it everywhere (incl. packet name, getter/setter names, Lua API method name, etc.)?

DS-Minetest replied:
What about `TOCLIENT_CAMERA_MODE_FLAGS`? Imho that the flags restrict seems kinda obvious.

SmallJoker replied:
`TOCLIENT_CAMERA_MODES` based on my yesterday's suggested design idea.

ClobberXD replied:
Yes, `TOCLIENT_CAMERA_MODES` is better suited for the name, due to SmallJoker's idea.

ClobberXD replied:
I thought of `TOCLIENT_CAMERA_MODES` as well, and that is also better suited for the packet name, due to [SmallJoker's idea](https://github.com/minetest/minetest/pull/8409#discussion_r270634673).

ClobberXD replied:
Shortened the name across the PR to camera modes. Packet renamed to `TOCLIENT_CAMERA_MODES` (yay!)

sofar commented:
neat idea, though, I like it.

nerzhul commented on src/client/camera.cpp:
```diff
@@ -602,6 +602,28 @@ void Camera::drawWieldedTool(irr::core::matrix4* translation)
 
+void Camera::toggleCameraMode()
+{
+	if (m_camera_mode == CAMERA_MODE_FIRST) {
```
switch case please

ClobberXD replied:
Done.

nerzhul commented on src/client/camera.h:
```diff
@@ -48,6 +48,10 @@ struct Nametag {
 };
 
+struct CameraModeRestrictions {
+	bool first, third, third_front;
```
initialize values in the definition please

ClobberXD replied:
Done; also added a ctor.

nerzhul commented on src/client/camera.h:
```diff
@@ -133,13 +137,16 @@ class Camera
-			m_camera_mode = CAMERA_MODE_FIRST;
+	void toggleCameraMode();
+
+	inline void setCameraModeRestrictions(CameraModeRestrictions &restrictions)
```
constness

ClobberXD replied:
Done. I hope that's what you asked for. :)

nerzhul commented on src/client/camera.h:
```diff
@@ -133,13 +137,16 @@ class Camera
+		m_camera_mode_restrictions = restrictions;
+	}
+
+	inline CameraModeRestrictions getCameraModeRestrictions() const
```
missing ref

ClobberXD replied:
Done.

nerzhul commented on src/network/clientpackethandler.cpp:
```diff
@@ -1474,3 +1475,15 @@ void Client::handleCommand_ModChannelSignal(NetworkPacket *pkt)
+	u16 byteflags;
+	*pkt >> byteflags;
+
+	CameraModeRestrictionFlags *flags = new CameraModeRestrictionFlags;
```
parenthesis please

ClobberXD replied:
Done. Also added a ctor for `CameraModeRestrictions`

nerzhul commented on src/client/camera.cpp:
```diff
@@ -60,6 +60,8 @@ Camera::Camera(MapDrawControl &draw_control, Client *client):
 	m_wieldnode->setItem(ItemStack(), m_client);
 	m_wieldnode->drop(); // m_wieldmgr grabbed it
 
+	m_camera_mode_restrictions = new CameraModeRestrictions();
```
memleak, this value is never delete. I suggest to use stack instead of pointer here, due to the usecase.

ClobberXD replied:
Done what I think you suggested. :)

nerzhul commented on src/client/camera.h:
```diff
@@ -48,6 +48,15 @@ struct Nametag {
+
+	CameraModeRestrictions()
+	{
+		first = third = third_front = false;
```
no, this is C++98 and lesser. This must be done on variable declaration.

```
bool first = false, third = false ...
```

ClobberXD replied:
Done

nerzhul commented on src/client/camera.h:
```diff
@@ -133,13 +142,18 @@ class Camera
+	void toggleCameraMode();
+
+	// Set camera mode restrictions
+	inline void setCameraModeRestrictions(const CameraModeRestrictions restrictions)
```
okay it's const, but where is the ref now ? :)

ClobberXD replied:
Done.

nerzhul commented on src/client/camera.h:
```diff
@@ -133,13 +142,18 @@ class Camera
+	}
+
+	// Get camera mode restrictions
+	inline CameraModeRestrictions &getCameraModeRestrictions() const
```
return const ref please

ClobberXD replied:
Done

nerzhul commented on src/player.h:
```diff
@@ -180,6 +180,7 @@ class Player
 	HudElement* removeHud(u32 id);
 	void        clearHud();
 
+	u16 camera_mode_restrictions;
```
initialize 0x01 | 0x02 | 0x04 here instead of the constructor function

ClobberXD replied:
Done.

nerzhul commented on src/client/camera.h:
```diff
@@ -176,10 +190,12 @@ class Camera
 
 	Client *m_client;
 
+	CameraModeRestrictions *m_camera_mode_restrictions;
```
this attribute must be private, we already have a setter and getter

ClobberXD replied:
It is private, as far as I can tell :)

nerzhul commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1806,6 +1806,36 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
+		return 0;
+
+	lua_newtable(L);
+	lua_pushboolean(L, player->camera_mode_restrictions & 0x01);
```
use the getter instead of the raw attribute please

ClobberXD replied:
I think the getter/setter methods are only accessible by the clients, and not servers - are files in `src/client/` compiled when building the server?

ClobberXD replied:
Nvm, added getter/setter methods for `RemotePlayer::m_camera_mode_restrictions`

ClobberXD commented:
Everything has been implemented, and this PR is ready for review. :tada: 

(This by no means indicates that the PR is functional, btw :P)

ClobberXD commented:
Trying to tackle this error:

```
/src/client/camera.cpp:43:1: error: uninitialized reference member in ‘struct CameraModeRestrictions&’
```

Let me know what's the best way to fix this. Should I not use references here? @nerzhul wanted its setter and getter to take and return references respectively. Shall I de-reference the passed reference param and store it in `m_camera_mode_restrictions`?

nerzhul commented on src/client/camera.h:
```diff
@@ -176,10 +187,12 @@ class Camera
 
 	Client *m_client;
 
+	CameraModeRestrictions &m_camera_mode_restrictions;
```
remove the reference here please it's just an owned object on the stack

ClobberXD replied:
Done

nerzhul commented on src/network/clientpackethandler.cpp:
```diff
@@ -1474,3 +1475,15 @@ void Client::handleCommand_ModChannelSignal(NetworkPacket *pkt)
+	u16 byteflags;
+	*pkt >> byteflags;
+
+	CameraModeRestrictions *flags = new CameraModeRestrictions();
```
don't use heap, just use stack

ClobberXD replied:
Done.

ClobberXD commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1806,6 +1806,54 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
+	u16 restrictions = player->getCameraModeRestrictions();
+
+	lua_newtable(L);
+	lua_pushboolean(L, restrictions & 0x01);
```
`camera.h` already contains the `CameraModes` enum, and I didn't want to duplicate the exact same enum elsewhere in the code. Should I move the existing enum to a common (i.e. accessible by both server and client) place, say, `player.h`?

SmallJoker replied:
I'd prefer to see something like this using `string_to_enum` and a `set` of the available camera modes.
```C++
const struct EnumString es_CameraModes[] =
{
	{CAMERA_MODE_FIRST,       "first"},
	{CAMERA_MODE_THIRD,       "third"}
	{CAMERA_MODE_THIRD_FRONT, "front"},
	{0, NULL},
};

std::set<CameraMode> m_camera_modes;

int ObjectRef::l_set_camera_modes(lua_State *L)
{
	// ...

	if (!lua_istable(L, 1))
		return 0; // must be like {}

	lua_pushnil(L);
	// Possibly 1, maybe some other index
	while (lua_next(L, 1) != 0) {
		int enum_val;
		if (string_to_enum(es_CameraModes, enum_val, lua_tostring(L, -2))) {
			if (readParam<bool>(L, -1, true))
				m_camera_modes.emplace((CameraMode)enum_val);
			else
				m_camera_modes.erase((CameraMode)enum_val);
		}
		lua_pop(L, 1);
	}

	// Needs at least one
	if (m_camera_modes.count() == 0)
		m_camera_modes.emplace(CAMERA_MODE_FIRST);
}

void Camera::toggleCameraMode()
{
	auto it = m_camera_modes.find(m_camera_mode);
	if (it == m_camera_modes.end()) {
		m_camera_mode = *m_camera_modes.begin();
		return;
	}

	if (++it == m_camera_modes.end())
		it = m_camera_modes.begin();
	m_camera_mode = *it;
}
```

ClobberXD replied:
Setting the camera modes themselves is much better than setting "restrictions" for the camera modes. :+1: 

Will do.

ClobberXD replied:
Please also suggest the best place to declare `std::set<CameraMode> m_camera_modes`. ;)

From your code above, I think it could be declared in `l_object.cpp`, but the problem is that `Camera::toggleCameraMode` won't be able to access it directly, as above.

How about in the `Player` class? Both the server and the client would store the camera modes in `RemotePlayer` and `LocalPlayer` respectively.

SmallJoker replied:
`Player` is probably the best choice. I'd use a getter function (non-const reference to the set) in this case to expose it to the Lua part.

ClobberXD replied:
Sounds good.

ClobberXD replied:
@SmallJoker How would these be sent over the network? Currently, these modes are encoded and sent as a `u16` over to the client, where the value is decoded again.

ClobberXD replied:
As I'm not familiar with `EnumString`, would you also be able to help me understand how the getter API method would work?

SmallJoker replied:
@ClobberXD You can either continue to send the entire vector (starting with length=`u16`) as `u16` (enum values) or `std::string` (Lua names). Both options would be OK.
Regarding the EnumString functions: 
 * Get enum pendant for a string value: `string_to_enum(enumstring_array, out_as_int, input_string)`
 * Get string pendant for an enum value: [TODO] `enum_to_string(enumstring_array, out_as_string, input_enum_as_int)` 

The idea of the code above would be that modders can indicate the changes of the available camera modes. If one mode is not specified in the setter function, it won't be updated.
The getter function would then push a table containing all camera modes: (iterate though all available modes until CAMERA_MODE_MAX (TODO).
```Lua
{
	first = true, -- contained in m_camera_modes
	third = false, -- not in m_camera_modes
	front = true
}
```

I see this is getting quite a long story for such a small feature. I'm sorry if this is tiring.
FYI: `luamethod(ObjectRef, set_camera_mode_restriction_flags),` needs updating in your source as well.

ClobberXD replied:
> You can either continue to send the entire vector (starting with length=u16) as u16 (enum values) or std::string (Lua names).

Alright, will look into it.

> Get string pendant for an enum value: **[TODO]** 
> `enum_to_string(enumstring_array, out_as_string, input_enum_as_int)`

Seeing `string_to_enum`, I already tried looking for `enum_to_string`, but couldn't find it.

> I see this is getting quite a long story for such a small feature. I'm sorry if this is tiring.

This is indeed getting too draggy, but I wholeheartedly support your approach as it's much more straightforward. The end result would be very robust and easy to work with, and that makes the wasted efforts worthwhile. Thanks for all your help. :)

>  luamethod(ObjectRef, set_camera_mode_restriction_flags), needs updating in your source as well.

Will do, thanks.

ClobberXD replied:
> The getter function would then push a table containing all camera modes: (iterate though all available modes until CAMERA_MODE_MAX (TODO)

I was wondering if there's a better way to iterate through all available modes. Maybe convert the set into a map (`std::map`)? Each player will own a copy of this map, and each key can be assigned `true` or `false`.

ClobberXD commented:
This PR finally compiles without errors, yay!

Tested the Lua API methods using the `luacmd` mod - the camera mode restrictions work flawlessly.

DS-Minetest commented on src/client/camera.cpp:
```diff
@@ -602,6 +602,32 @@ void Camera::drawWieldedTool(irr::core::matrix4* translation)
+				m_camera_mode = CAMERA_MODE_FIRST;
+			else if (!m_camera_mode_restrictions.third)
+				m_camera_mode = CAMERA_MODE_THIRD;
+		}
```
A break is missing here. (https://dev.minetest.net/Code_style_guidelines#Miscellaneous)

Also there are many curly brackets. I don't know whether it looks weird to make one block per case.

And everything is indented too much (see https://www.kernel.org/doc/html/latest/process/coding-style.html#indentation).

ClobberXD replied:
Thanks for pointing that out. Fixed. :+1:

ClobberXD commented:
Can't take it anymore; squashing... :)

ClobberXD commented:
@SmallJoker I think this PR is ready. I've added documentation for the new Lua API methods. I've also implemented `enum_to_string`, but now I don't know where I should use it, lol :)

I still haven't tested this yet (I haven't even compiled it) - I'll do it when I get some time later today.

rubenwardy commented:
APIs like this make me feel uncomfortable. I would quite like to have a unified way to control the current state a user is in in and which states they can change to. Unified in Lua API, packets, and handling code. The problem with hud flags is that they specifically are for HUD, and they are also readonly by the client. Ie: it's a flag, not a command to go into a particular state.

Argh, ramble. But maybe something like this would be better:

```c++
player:set_flags({
	camera_mode_first = true,
	minimap = true,
})

player:set_camera_mode("first")
player:se
-- all these methods send the same packets
```

obviously changing anything over than camera mode related things are out of scope, but it would be nice to have a plan for a unified API and packet handling

ClobberXD commented:
I really agree that we need a more unified API for all these features. We already have too many packets going to and fro, and there are separate getters and setters for each and every feature, making the API cluttered. While it's impractical to implement the unified API, I'm open to suggestions for incorporating a more unified approach into this feature PR.

ClobberXD commented:
@nerzhul Would you be able to update (or dismiss) your review?

paramat commented:
I dismissed nerzhul's review, i assume it was all attended to.

@nerzhul i have a request, please could you dismiss your reviews when the issues are resolved? You never seem to and i have had clean those up very many times.
If you don't want the bother of having to dismiss them (i can understand this) then you can use the 'add single comment' option instead of the review option, when leaving line comments.

Also , when you +1 a PR please add the appropriate 'approval' label.

ClobberXD commented:
Rebased; will test again. But this PR is complete, and only requires reviews.

paramat commented:
Code still broken?

ClobberXD commented:
@paramat Yes, I wasn't able to fix it within the evening. It seems fixable though, nothing to worry. :)

paramat commented:
No problem, just needed to know to re-add WIP label.

ClobberXD commented:
Added `cout`s everywhere. I noticed that the server completely freezes (client still works fine, even in singleplayer) when executing `Server::SendCameraModes`. This is the function that serialises the camera mode enums and sends them over the network. It's a very simple and straightforward function, and I don't see anything wrong with it. Will dig deeper...

ClobberXD commented:
What should be the default camera modes of players, if the server doesn't immediately send camera modes upon join? The most restrictive, with only first-person mode? Or the least restrictive, with all three modes? I personally prefer the most restrictive option, as I believe that the server should have control over this aspect, but the least restrictive is the backwards-compatible approach.

paramat commented on doc/lua_api.txt:
```diff
@@ -5881,6 +5881,18 @@ This is basically a reference to a C++ `ServerActiveObject`
     * Returns `false` if failed.
     * Resource intensive - use sparsely
     * To get blockpos, integer divide pos by 16
+* `set_camera_modes(modes)`: Sets the camera modes for the player.
```
"Sets the allowed camera modes ..."

paramat commented:
Hmm good question, difficult to decide.
Allowing all 3 modes is harmless in most situations, and certainly harmless for creative servers.
If the default is first only, considering that most servers will just leave the default in effect, this would then be very unpopular and unnecessarily restrictive.
So in this case i think i prefer allowing all modes by default.

ClobberXD commented:
> If the default is first only, considering that most servers will just leave the default in effect, this would then be very unpopular and unnecessarily restrictive.

Good point, but this is more of a game/mod thing - server owners can make custom mods if they want, but it's mostly upto games like MTG to decide upon this (this would be set by the player api mod, I think?). But then again, games/mods would be *forced* to implement this, and even if they do end up adding the necessary code, it'd mean that all servers would have to update to the latest versions.

So I think the least restrictive option is the way to go (though I personally prefer the latter), for the sake of backwards-compat.

ClobberXD commented:
Rebased. I'll continue to debug the server freeze now.

esanhueza commented:
Hi, new here. I took the work made by @ClobberXD, added some corrections and it's working just fine over the master branch. But I need to solve two things.
1. If the player is in first person mode and a Mod restrict the camera modes to allow only third person mode, should I change the camera mode of the player immediately? or should I let that work for the Mod?
2. I don't know how to share my changes. As I understand, I have to make a pull request, but this feature already have one. How should I proceed?
